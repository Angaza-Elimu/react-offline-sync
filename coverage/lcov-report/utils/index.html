
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">73.13% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>49/67</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">47.06% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>16/34</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">61.11% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>11/18</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">74.19% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>46/62</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="checkConnectivityInterval.ts"><a href="checkConnectivityInterval.ts.html">checkConnectivityInterval.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file low" data-value="checkInternetAccess.ts"><a href="checkInternetAccess.ts.html">checkInternetAccess.ts</a></td>
	<td data-value="21.43" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 21%"></div><div class="cover-empty" style="width: 79%"></div></div>
	</td>
	<td data-value="21.43" class="pct low">21.43%</td>
	<td data-value="14" class="abs low">3/14</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="23.08" class="pct low">23.08%</td>
	<td data-value="13" class="abs low">3/13</td>
	</tr>

<tr>
	<td class="file high" data-value="constants.ts"><a href="constants.ts.html">constants.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file medium" data-value="getSimilarActionInQueue.ts"><a href="getSimilarActionInQueue.ts.html">getSimilarActionInQueue.ts</a></td>
	<td data-value="50" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 50%"></div><div class="cover-empty" style="width: 50%"></div></div>
	</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="8" class="abs medium">4/8</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="4" class="abs low">1/4</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="6" class="abs medium">3/6</td>
	</tr>

<tr>
	<td class="file high" data-value="makeHttpRequest.ts"><a href="makeHttpRequest.ts.html">makeHttpRequest.ts</a></td>
	<td data-value="92.31" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 92%"></div><div class="cover-empty" style="width: 8%"></div></div>
	</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="26" class="abs high">24/26</td>
	<td data-value="64.29" class="pct medium">64.29%</td>
	<td data-value="14" class="abs medium">9/14</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="26" class="abs high">24/26</td>
	</tr>

<tr>
	<td class="file low" data-value="nonNullable.ts"><a href="nonNullable.ts.html">nonNullable.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	</tr>

<tr>
	<td class="file high" data-value="wait.ts"><a href="wait.ts.html">wait.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Tue Jun 22 2021 20:09:59 GMT+0300 (East Africa Time)
            </div>
        </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    