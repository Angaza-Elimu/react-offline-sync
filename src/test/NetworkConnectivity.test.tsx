import * as React from 'react';
import NetInfo, { NetInfoStateType } from '@react-native-community/netinfo';
import { AppState, Platform, View } from 'react-native';
import { shallow } from 'enzyme';
import { render } from 'react-native-testing-library';
import NetworkConnectivity, {
  RequiredProps,
} from '../components/NetworkConnectivity';
import { clear, setup } from '../utils/checkConnectivityInterval';
import checkInternetAccess from '../utils/checkInternetAccess';
import entries from '../utils/objectEntries';
import { DEFAULT_CUSTOM_HEADERS } from '../utils/constants';

type OptionalProps = Omit<RequiredProps, 'children'>;
type GetElementParams<P = any> = {
  props?: Pick<RequiredProps, 'children'> & Partial<OptionalProps>;
  Component?: React.ComponentType<P>;
};

interface MethodsMap {
  getConnectionChangeHandler?: any;
  intervalHandler?: any;
  setState?: any;
}
const mockConnectionChangeHandler = jest.fn();
const mockGetConnectionChangeHandler = jest.fn(
  () => mockConnectionChangeHandler,
);
const mockIntervalHandler = jest.fn();
const mockHandleNetInfoChange = jest.fn();
const mockHandleConnectivityChange = jest.fn();
const mockCheckInternet = jest.fn();

const addEventListener = NetInfo.addEventListener as jest.Mock;
const unsubscribe = jest.fn();
const fetch = NetInfo.fetch as jest.Mock;
jest.mock('../utils/checkConnectivityInterval');
jest.mock('../utils/checkInternetAccess', () =>
  jest.fn().mockResolvedValue(true),
);

/**
 * Helper function that creates a class that extends NetworkConnectivity
 * and mocks the specific methods on the prototype,
 * in order to not affect the rest of the tests
 * @param methodsMap
 * @returns {ClassWithMocks}
 */

function mockPrototypeMethods(methodsMap: MethodsMap = {} as MethodsMap) {
  class ClassWithMocks extends NetworkConnectivity {}
  entries(methodsMap).forEach(
    ([method, mockFn]) => (ClassWithMocks.prototype[method] = mockFn),
  );
  return ClassWithMocks;
}

const ChildrenComponent = () => <View />;

const initialProps = {
  children: ChildrenComponent,
};

const getElement = ({
  props = initialProps,
  Component = NetworkConnectivity,
}: GetElementParams = {}) => {
  const { children, ...rest } = props;
  return <Component {...rest}>{children}</Component>;
};

describe('NetworkConnectivity', () => {
  afterEach(() => {
    addEventListener.mockClear();
    fetch.mockClear();
    mockConnectionChangeHandler.mockClear();
    mockGetConnectionChangeHandler.mockClear();
    mockIntervalHandler.mockClear();
    mockHandleNetInfoChange.mockClear();
    mockHandleConnectivityChange.mockClear();
    mockCheckInternet.mockClear();
  });

  it('defaultProps', () => {
    expect(NetworkConnectivity.defaultProps).toMatchSnapshot();
  });

  it('passes the connection state into the FACC', () => {
    const children = jest.fn();
    shallow(getElement({ props: { children } }));
    expect(children).toHaveBeenCalledWith({ isConnected: true });
  });

  describe('componentDidMount', () => {
    describe('iOS, pingInterval = 0', () => {
      it(`sets up a NetInfo.isConnected listener for connectionChange 
      AND does NOT call setupConnectivityCheckInterval`, () => {
        Platform.OS = 'ios';
        const MockedNetworkConnectivity = mockPrototypeMethods({
          getConnectionChangeHandler: mockGetConnectionChangeHandler,
        });
        shallow(
          getElement({
            Component: MockedNetworkConnectivity,
          }),
        );
        expect(NetInfo.addEventListener).toHaveBeenCalledTimes(1);
        expect(NetInfo.addEventListener).toHaveBeenCalledWith(
          mockConnectionChangeHandler,
        );
        expect(setup).not.toHaveBeenCalled();
      });
    });


    it(`calls setupConnectivityCheckInterval with the right arguments
     WHEN pingInterval is higher than 0`, () => {
      Platform.OS = 'ios';
      const MockedNetworkConnectivity = mockPrototypeMethods({
        intervalHandler: mockIntervalHandler,
      });
      shallow(
        getElement({
          Component: MockedNetworkConnectivity,
          props: {
            children: ChildrenComponent,
            pingInterval: 1000,
          },
        }),
      );
      expect(setup).toHaveBeenCalled();
    });
  });

 

  describe('getConnectionChangeHandler', () => {
    it('returns this.handleNetInfoChange when props.shouldPing = true', () => {
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props: {
            children: ChildrenComponent,
            shouldPing: true,
          },
        }),
      );
      wrapper.instance().handleNetInfoChange = mockHandleNetInfoChange;
      expect(wrapper.instance().getConnectionChangeHandler()).toBe(
        mockHandleNetInfoChange,
      );
    });

    it('returns this.handleConnectivityChange when props.shouldPing = false', () => {
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props: {
            children: ChildrenComponent,
            shouldPing: false,
          },
        }),
      );
      wrapper.instance().handleConnectivityChange = mockHandleConnectivityChange;
      expect(wrapper.instance().getConnectionChangeHandler()).toBe(
        mockHandleConnectivityChange,
      );
    });
  });

  describe('checkInternet', () => {
    it('returns early if pingIfBackground = false AND app is not in the foreground', async () => {
      AppState.currentState = 'background';
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props: {
            children: ChildrenComponent,
            pingInBackground: false,
          },
        }),
      );
      wrapper.instance().handleConnectivityChange = mockHandleConnectivityChange;
      await wrapper.instance().checkInternet();
      expect(checkInternetAccess).not.toHaveBeenCalled();
      expect(mockHandleConnectivityChange).not.toHaveBeenCalled();
    });

    it(`calls checkInternetAccess AND handleConnectivityChange 
    with the right arguments if app is in foreground`, async () => {
      const props = {
        pingTimeout: 2000,
        pingServerUrl: 'dummy.com',
        httpMethod: 'OPTIONS' as 'OPTIONS',
        children: ChildrenComponent,
        customHeaders: DEFAULT_CUSTOM_HEADERS,
      };
      AppState.currentState = 'active';
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props,
        }),
      );
      wrapper.instance().handleConnectivityChange = mockHandleConnectivityChange;
      await wrapper.instance().checkInternet();
      expect(checkInternetAccess).toHaveBeenCalledWith({
        url: props.pingServerUrl,
        timeout: props.pingTimeout,
        method: props.httpMethod,
        customHeaders: props.customHeaders,
      });
      expect(mockHandleConnectivityChange).toHaveBeenCalledWith({
        isConnected: true,
      });
    });
  });

  describe('intervalHandler', () => {
    it('returns early if there is connection AND pingOnlyIfOffline = true', () => {
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props: {
            children: ChildrenComponent,
            pingOnlyIfOffline: true,
          },
        }),
      );
      wrapper.instance().checkInternet = mockCheckInternet;
      wrapper.setState({ isConnected: true });
      wrapper.instance().intervalHandler();
      expect(mockCheckInternet).not.toHaveBeenCalled();
    });

    it(`calls checkInternet if it's not connected OR pingOnlyIfOffline = false`, () => {
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          props: {
            children: ChildrenComponent,
            pingOnlyIfOffline: false,
          },
        }),
      );
      wrapper.instance().checkInternet = mockCheckInternet;
      wrapper.setState({ isConnected: false });
      wrapper.instance().intervalHandler();
      expect(mockCheckInternet).toHaveBeenCalledTimes(1);

      wrapper.setState({ isConnected: true });
      wrapper.instance().intervalHandler();
      expect(mockCheckInternet).toHaveBeenCalledTimes(2);
    });
  });

  describe('handleConnectivityChange', () => {
    it('calls setState with the new connection value', () => {
      const mockSetState = jest.fn();
      const MockedNetworkConnectivity = mockPrototypeMethods({
        setState: mockSetState,
      });
      const wrapper = shallow<NetworkConnectivity>(
        getElement({
          Component: MockedNetworkConnectivity,
        }),
      );

      wrapper.instance().handleConnectivityChange({
        type: 'other' as NetInfoStateType.other,
        isConnected: true,
        isInternetReachable: true,
        details: {
          cellularGeneration: null,
          isConnectionExpensive: false,
        },
      });
      expect(mockSetState).toHaveBeenCalledWith({ isConnected: true });

      wrapper.instance().handleConnectivityChange({
        type: 'none' as NetInfoStateType.none,
        isConnected: false,
        isInternetReachable: false,
        details: null,
      });
      expect(mockSetState).toHaveBeenCalledWith({ isConnected: false });
    });
  });


});
