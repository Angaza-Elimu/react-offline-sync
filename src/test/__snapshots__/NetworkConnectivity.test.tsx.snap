// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkConnectivity defaultProps 1`] = `
Object {
  "customHeaders": Object {},
  "httpMethod": "HEAD",
  "onConnectivityChange": [Function],
  "pingInBackground": false,
  "pingInterval": 0,
  "pingOnlyIfOffline": false,
  "pingServerUrl": "https://www.google.com/",
  "pingTimeout": 10000,
  "shouldPing": true,
}
`;
